scipy/optimize/_qap.py

```diff
@@ -0,0 +1,408 @@
+    perform the following calculation.
+
+    >>> score = np.trace(cost.T @ dist[np.ix_(res['col_ind'], res['col_ind'])])
+    >>> print(score)
+    3260
+
+    Here, we are simply permuting the distance matrix.
+
+    """
+
+    if options is None:
+        options = {}
+
+    return _quadratic_assignment_faq(cost_matrix, dist_matrix, maximize,
+                                     **options)
+
+
+def _quadratic_assignment_faq(
+        cost_matrix,
+        dist_matrix,
+        maximize=False,
+        partial_match=None,
+        init="barycenter",
+        init_weight=None,
+        init_n=1,
+        maxiter=30,
+        shuffle_input=True,
+        eps=0.05
+):
+
+    cost_matrix = np.asarray(cost_matrix)
+    dist_matrix = np.asarray(dist_matrix)
+
+    if partial_match is None:
+        partial_match = np.array([[], []]).T
+    partial_match = np.asarray(partial_match)
+    init_n = operator.index(init_n)
+    maxiter = operator.index(maxiter)
+
+    # ValueError check
+    msg = None
+    if cost_matrix.shape[0] != cost_matrix.shape[1]:
+        msg = "'cost_matrix' must be square"
+    elif dist_matrix.shape[0] != dist_matrix.shape[1]:
+        msg = "'dist_matrix' must be square"
+    elif cost_matrix.shape != dist_matrix.shape:
+        msg = "Adjacency matrices must be of equal size"
+    elif (cost_matrix < 0).any() or (dist_matrix < 0).any():
+        msg = "Adjacency matrix contains negative entries"
+    elif partial_match.shape[0] > cost_matrix.shape[0]:
+        msg = "There cannot be more seeds than there are nodes"
+    elif partial_match.shape[1] != 2:
+        msg = "`partial_match` must have two columns"
+    elif (partial_match < 0).any():
+        msg = "`partial_match` contains negative entries"
+    elif (partial_match >= len(cost_matrix)).any():
+        msg = "`partial_match` entries must be less than the number of nodes"
+    elif not len(set(partial_match[:, 0])) == len(partial_match[:, 0]) or not \
+            len(set(partial_match[:, 1])) == len(partial_match[:, 1]):
+        msg = "`partial_match` column entries must be unique"
+    elif isinstance(init, str) and init not in {'barycenter'}:
+        msg = "Invalid 'init' parameter string"
+    elif init_weight is not None and (init_weight < 0 or init_weight > 1):
+        msg = "'init_weight' must be in range [0, 1]"
+    elif init_n <= 0:
+        msg = "'n_init' must be a positive integer"
+    elif maxiter <= 0:
+        msg = "'maxiter' must be a positive integer"
+    if msg is not None:
+        raise ValueError(msg)
+
+    # TypeError check
+    if not isinstance(shuffle_input, bool):
+        msg = "'shuffle_input' must be a boolean"
+    elif eps <= 0 or type(eps) is not float:
+        msg = "'eps' must be a positive float"
+    elif not isinstance(maximize, bool):
+        msg = "'maximize' must be a boolean"
+    if msg is not None:
+        raise TypeError(msg)
+
+    rng = np.random.RandomState()
```

Comment:
There needs to be a way for the user to control this, because they need to be able to get the same result repeatedly.